### PR TITLE
Add ConvFit temperature correction function

### LINKED ISSUE #28891: ConvFit temperature correction does not work with negative values
Currently the temperature correction on the ConvFit tab does not work with negative energy values.

### PR DESCRIPTION
**Pull request by StephenSmith25**
**Description of work.**
This PR creates a temperature correction function which will be used in the ConvFit tab in place of the user defined exponential. This is necessary as the user defined exponential is undefined at Q=0, which means the direct convolution of a function using this correction results in errors (due to NaNs). This function prevents the undefined behavior by using f(Q)=1 when Q is zero. This is the limit of the original temperature correction function as Q->0. 

**To test:**

- Data to test each tab of the Indirect Data Analysis GUI can be found in #26631.
- Open the Indirect Data Analysis GUI
- Go to the ConvFit tab
- Load in data and a resolution file
- Add a fit type (e.g Lorentzian)
- Add a temperature correction with the checkbox in the browser (any will do).
- The previous issue was encountered when the fitting range was changed to be non-symmetric (different number of positive and negative points). Change the range to approximately Start X = -0.163169 and End X = 0.104346.
- Change the fit spectra range to `0 to 1`
- Do a fit, it should now work

 

Fixes #28891 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by cailafinn — DISMISSED** [review]
Fixes the described issue and definitely neatens up the temperature correction code. Good work 👍

File: Framework/CurveFitting/inc/MantidCurveFitting/Functions/ConvTempCorrection.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+public:
+  std::string name() const override { return "ConvTempCorrection"; }
+  const std::string category() const override { return "QuasiElastic"; }
+
+protected:
+  void function1D(double *out, const double *xValues,
+                  const size_t nData) const override;
+  void init() override;
+
+private:
```
Comment (DavidFair) [inline Framework/CurveFitting/inc/MantidCurveFitting/Functions/ConvTempCorrection.h:30]: Don't worry about changing this, but you can omit any labels such as private if your not using them

File: Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#include "MantidAPI/FunctionFactory.h"
+#include <cmath>
+
+namespace {
+constexpr double CONVERSIONFACTOR = 11.606;
+} // namespace
+
+namespace Mantid {
+namespace CurveFitting {
+namespace Functions {
```
Comment (DavidFair) [inline Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp:18]: Also again, don't worry about changing this, but you can now do this instead with C++17:
`namespace Mantid::CurveFitting::Functions {`

File: Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+
+#include "MantidCurveFitting/Functions/ConvTempCorrection.h"
+#include "MantidAPI/FunctionFactory.h"
+#include <cmath>
+
+namespace {
+constexpr double CONVERSIONFACTOR = 11.606;
```
Comment (DavidFair) [inline Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp:13]: We need a comment explaining what this value is / how it was derived or a citation

File: Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+void ConvTempCorrection::function1D(double *out, const double *xValues,
+                                    const size_t nData) const {
+  const double T = getParameter("Temperature");
+  for (size_t i = 0; i < nData; ++i) {
+    double x = xValues[i];
+    if (x == 0.0) {
+      out[i] = 1.00;
+    } else {
+      out[i] =
+          (x * CONVERSIONFACTOR) / T / (1 - exp(-((x * CONVERSIONFACTOR) / T)));
```
Comment (DavidFair) [inline Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp:35]: Ditto on a comment or citation on this formula, what is it?

File: Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+} // namespace
+
+namespace Mantid {
+namespace CurveFitting {
+namespace Functions {
+
+DECLARE_FUNCTION(ConvTempCorrection)
+
+void ConvTempCorrection::init() {
+  declareParameter("Temperature", 300.0, "Temperature correction value");
```
Comment (DavidFair) [inline Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp:23]: Could we put a unit in the doc string, I'm assuming this is Kelvin

**Review by DavidFair — CHANGES_REQUESTED** [review]

File: Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp (hunk 3 above)
Comment (StephenSmith25) [reply Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp:13]: Ill try get hold of some sort of reference tomorrow, I was mostly just copy and pasting from the original declaration, but there must be a source somewhere.

File: Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp (hunk 4 above)
Comment (StephenSmith25) [reply Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp:35]: @SpencerHowells Do you know the source of this temperature correction function?

File: Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp (hunk 3 above)
Comment (AnthonyLim23) [reply Framework/CurveFitting/src/Functions/ConvTempCorrection.cpp:13]: its probably 1/k  in funny units (e.g. 0.695 cm^{-1}/K)

**Comment by AnthonyLim23** [comment]
if your adding a function you should add a unit test and documentation

**Review by DavidFair — APPROVED** [review]
Retested following the changes, everything still works
Thanks for digging out citations for these conversions  - I'm sure it will make a future developer or two happy

**Pull request merged**